fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java

```diff
@@ -242,37 +276,89 @@ public int compare(DelinquencyRange o1, DelinquencyRange o2) {
         }
     }
 
-    private boolean isOverlaped(DelinquencyRange o1, DelinquencyRange o2) {
-        return Math.max(o1.getMinimumAgeDays(), o2.getMinimumAgeDays()) <= Math.min(o1.getMaximumAgeDays(), o2.getMaximumAgeDays());
+    private boolean isOverlapped(DelinquencyRange o1, DelinquencyRange o2) {
+        if (o2.getMaximumAgeDays() != null) { // Max Age undefined - Last one
+            return Math.max(o1.getMinimumAgeDays(), o2.getMinimumAgeDays()) <= Math.min(o1.getMaximumAgeDays(), o2.getMaximumAgeDays());
+        } else {
+            return Math.max(o1.getMinimumAgeDays(), o2.getMinimumAgeDays()) <= Math.min(o1.getMaximumAgeDays(), o2.getMinimumAgeDays());
+        }
+    }
+
+    private Map<String, Object> lookUpDelinquencyRange(final Loan loan, final DelinquencyBucket delinquencyBucket, long ageDays) {
+        Map<String, Object> changes = new HashMap<>();
+
+        if (ageDays <= 0) { // No Delinquency
+            log.info("Loan {} without delinquency range with {} days", loan.getId(), ageDays);
+            changes = setLoanDelinquencyTag(loan, null);
+
+        } else {
+            for (DelinquencyRange delinquencyRange : delinquencyBucket.getRanges()) {
+                if (delinquencyRange.getMaximumAgeDays() == null) {
+                    if (delinquencyRange.getMinimumAgeDays() <= ageDays) {
+                        log.info("Loan {} with delinquency range {} with {} days", loan.getId(), delinquencyRange.getClassification(),
+                                ageDays);
+                        changes = setLoanDelinquencyTag(loan, delinquencyRange.getId());
+                        break;
```

Comment:
Since you are breaking out of the loop and the way the logic is implemented, I think the delinquency bucket ranges should be an ordered list, am I wrong?
In the current case, sometimes it might be an ordered list but sometimes it might not be unless we do an explicit ordering based on the ages.
Thoughts?

Reply:
I've updated that, but honestly I don't think that it can add some value, because the range are not overlapped, the break is more to avoid the range validation of the rest ranges once time that you have find the right range